packages/compiler-cli/src/ngtsc/imports/src/emitter.ts

```diff
@@ -233,6 +238,28 @@ export class LogicalProjectStrategy implements ReferenceEmitStrategy {
   }
 }
 
+/**
+ * A `ReferenceEmitStrategy` which constructs relatives paths between `ts.SourceFile`s.
+ *
+ * This strategy can be used if there is no `rootDir`/`rootDirs` structure for the project which
+ * necessitates the stronger logic of `LogicalProjectStrategy`.
+ */
+export class RelativePathStrategy implements ReferenceEmitStrategy {
+  constructor(private reflector: ReflectionHost) {}
+
+  emit(ref: Reference<ts.Node>, context: ts.SourceFile): Expression|null {
+    const destSf = getSourceFile(ref.node);
+    let moduleName =
+        relative(dirname(resolve(context.fileName)), resolve(stripExtension(destSf.fileName)));
```

Comment:
nit: I would move the `stripExtension` to be last, and maybe use `absoluteFromSourceFile`:
```suggestion
        stripExtension(relative(dirname(absoluteFromSourceFile(context)), absoluteFromSourceFile(destSf)));
```

Reply:
`stripExtension` is `(string) => string`, and everything else deals in `AbsoluteFsPath`, so it makes sense as the inner operation.

But I do like the idea of `absoluteFromSourceFile` - doing that.

Reply:
It took me a moment to understand this, but 👍